_posts/2024-02-09-Database-interaction-with-skunk-scala-library:-A-comprehensive-guide.md

```diff
@@ -0,0 +1,555 @@
+          cmd.execute(userData).flatMap { rowCount =>
+            IO(println(s"Inserted $rowCount row(s)"))
+          }
+        } *> IO.pure(ExitCode.Success)
+      }
+```
+After calling `Session.prepare`, a prepared command is created using our defined command to be executed on `userData`. Following this, the `execute` method is used, and the number of affected rows is returned. Finally, we print the user-created ID to the console. When this command is executed, the output will be similar to the following.
+```
+Inserted Insert(1) row(s) with id: b68f2676-611e-4db2-af80-0458e9c52bd3
+```
+To fetch the user we created using the obtained id, we will use a skunk query. The input of the query will be the user's id, and the output will be a User object. The query is constructed as follows:
+```scala
+    ....
+    ....
+      val selectById: Query[UUID, User] =
+        sql"""
+            SELECT * FROM users
+            WHERE id = $uuid
+          """.query(codec)
+```
+It's worth noting that we still use `codec` conversion to automatically convert obtained PostgreSQL types to our Scala types. The query to be executed is prepared similarly to the previous command.
+```scala
+    ....
+    ....
+      resource.use { session =>
+        session.prepare(selectById).flatMap { pq =>
+          pq.option(UUID.fromString("b68f2676-611e-4db2-af80-0458e9c52bd3")).flatMap { res =>
+            res match
+              case Some(user) => IO(println(s"User found: $user"))
+              case None       => IO(println(s"No user found with id: $userId")) *> IO.raiseError(new NoSuchElementException())
+          }
+        } *> IO.pure(ExitCode.Success)
+```
+After preparing our query, we fetch a maximum of one row by using the `option` method along with the provided id as an argument. If a user is found, we'll print the result to the console. However, if no user is found, we'll raise an error with a message indicating that no user was found. Executing the query with a known user id should give an output similar to the following
+```
+User found: User(b68f2676-611e-4db2-af80-0458e9c52bd3,Jacob,jacob@email.com)
+```
+
+
+## 9. Refactoring our code
+Let's refactor our code in a functional style and add support for basic CRUD operations.
+Given we're leveraging the use of a purely functional library, we shall continue writing purely functional code throughout using the tagless final approach.
+
+### 9.1. Database Repository
+We define a trait called `Repository` to handle our queries and commands for interacting with our database. This file is placed in `src/main/scala/db`.
+```scala
+    import cats.effect.{Resource, Sync}
+    import cats.implicits.*
+    import skunk_guide.errors.Errors.UniqueViolation
+    import skunk.{Command, Query, Session, SqlState}
+
+    // While F is the effect type, E represents the domain/model used to perform operations
+    trait Repository[F[_], E](resource: Resource[F, Session[F]]) {
```

Comment:
Don't pass a resource itself as an argument, because the resource will be evaluated (and the Session recreated) every time you call the `run` method. We need to pass the `Session[F]` instance. After that, we won't need the `run` method anymore.

Another thing that I don't understand is the type argument E - isn't that supposed to be the same as the type A of the methods below?

Reply:
`E` represents the domain or data type to be used by the repository e.g. `User`. While `A` represents the argument type passed to a `Query` or a `Command`.